lms/djangoapps/grades/tests/test_models.py

```diff
@@ -0,0 +1,199 @@
+        repeat_vblocks = VisibleBlocks.objects.create_from_blockrecords([self.record_b, self.record_a])
+        new_vblocks = VisibleBlocks.objects.create_from_blockrecords([self.record_b])
+
+        self.assertEqual(stored_vblocks.pk, repeat_vblocks.pk)
+        self.assertEqual(stored_vblocks.hashed, repeat_vblocks.hashed)
+
+        self.assertNotEqual(stored_vblocks.pk, new_vblocks.pk)
+        self.assertNotEqual(stored_vblocks.hashed, new_vblocks.hashed)
+
+    def test_blocks(self):
+        """
+        Ensures that, given an array of BlockRecord, creating visible_blocks and accessing visible_blocks.blocks yields
+        a copy of the initial array. Also, trying to set the blocks property should raise an exception.
+        """
+        expected_blocks = [self.record_a, self.record_b]
+        visible_blocks = VisibleBlocks.objects.create_from_blockrecords(expected_blocks)
+        self.assertEqual(BlockRecordSet(expected_blocks), visible_blocks.blocks)
+        with self.assertRaises(AttributeError):
+            visible_blocks.blocks = expected_blocks
+
+
+@ddt.ddt
+class PersistentSubsectionGradeTest(GradesModelTestCase):
+    """
+    Test the PersistentSubsectionGrade model.
+    """
+    def setUp(self):
+        super(PersistentSubsectionGradeTest, self).setUp()
+        self.usage_key = BlockUsageLocator(
+            course_key=self.course_key,
+            block_type='subsection',
+            block_id='subsection_12345',
+        )
+        self.params = {
+            "user_id": 12345,
+            "usage_key": self.usage_key,
+            "course_version": "deadbeef",
+            "subtree_edited_date": "2016-08-01 18:53:24.354741",
+            "earned_all": 6,
+            "possible_all": 12,
+            "earned_graded": 6,
+            "possible_graded": 8,
+            "visible_blocks": [self.record_a, self.record_b],
+        }
+
+    def test_create(self):
+        """
+        Tests model creation, and confirms error when trying to recreate model.
+        """
+        created_grade = PersistentSubsectionGrade.objects.create(**self.params)
+        read_grade = PersistentSubsectionGrade.read_grade(
+            user_id=self.params["user_id"],
+            usage_key=self.params["usage_key"],
+        )
+        self.assertEqual(created_grade, read_grade)
+        with self.assertRaises(ValidationError):
+            created_grade = PersistentSubsectionGrade.objects.create(**self.params)
+
+    def test_create_bad_params(self):
+        """
+        Confirms create will fail if params are missing.
+        """
+        del self.params["earned_graded"]
+        with self.assertRaises(ValidationError):
+            PersistentSubsectionGrade.objects.create(**self.params)
+
+    def test_course_version_is_optional(self):
+        del self.params["course_version"]
+        PersistentSubsectionGrade.objects.create(**self.params)
```

Comment:
Assert that the value of `course_version` is `None` when the grade record is retrieved.